app/models/compute_resources/foreman/model/vmware.rb

```diff
@@ -488,8 +497,22 @@ def parse_args(args)
 
       args.except!(:hardware_version) if args[:hardware_version] == 'Default'
 
-      firmware_type = args.delete(:firmware_type)
-      args[:firmware] = firmware_mapping(firmware_type) if args[:firmware] == 'automatic'
+      firmware_type = args.delete(:firmware_type).to_s
+
+      if args[:firmware] == 'automatic'
+        if firmware_type == 'none' || firmware_type.empty?
+          args[:firmware] = 'bios'
+        else
+          args[:firmware] = firmware_type
+        end
+      end
+
+      if args[:firmware]&.start_with?('uefi')
+        args[:secure_boot] = true if args[:firmware] == 'uefi_secure_boot'
+        args[:firmware] = 'efi'
+      end
+
+      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args.delete(:virtual_tpm))
```

Comment:
No need to delete it:
```suggestion
      args[:virtual_tpm] = ActiveRecord::Type::Boolean.new.cast(args[:virtual_tpm])
```

I'd still insist that we raise an exception here. That way `clone_vm` (which also calls parse_args) can benefit from it as well.

When we earlier discussed adding an error, it was important to also add a return if there are any errors (like `create_vm` already does).

The benefit of `errors` is that you can actually flag the field itself that is wrong, instead of some generic unspecified exception.

Reply:
I believe `clone_vm` shouldn’t encounter this issue. Based on my review, it also calls the `create_vm` method later in the process, and the workflow is consistent with that for creating a new VM.

Regarding adding this as an error, it might not be very effective since we would still end up initializing the Fog::Vsphere::Compute::Server object [here](https://github.com/theforeman/foreman/blob/develop/app/models/compute_resources/foreman/model/vmware.rb#L552).

IMO, the correct place to raise this exception is in the `parse_args` method. However, since we don't have the validations for this, this would end up in a 500 error code, causing the whole page to break. 
Unfortunately, I couldn't find a better solution for this (that doesn't require refactoring this entire part of the code).
As I mentioned in my previous comment, we lack this functionality for all the fields sent to Fog::Vsphere. Addressing it for just this specific case will not be sufficient.